pkg/slayers/path/scion/raw_spec.gobra

```diff
@@ -666,4 +667,65 @@ func (s *Raw) LastHopLemma(ubuf []byte, dp io.DataPlaneSpec) {
 	pkt := reveal s.absPkt(dp, ubuf)
 	assert pkt.CurrSeg == reveal CurrSeg(ubuf, offset+prevSegLen, currINFIdx, currHFIdx-prevSegLen, segLen, 0)
 	assert len(pkt.CurrSeg.Future) == 1
-}
\ No newline at end of file
+}
+
+ghost
+opaque
+requires 0 <= start && start <= middle
+requires acc(raw.Mem(ubuf), R50)
+requires acc(sl.AbsSlice_Bytes(ubuf, 0, len(ubuf)), R50)
+requires unfolding acc(raw.Mem(ubuf), R50) in middle + path.HopLen <= end && end <= len(raw.Raw)
+decreases
+pure func (raw *Raw) CorrectlyDecodedHF_Raw(h path.HopField, start int, middle int, end int, ubuf []byte) bool {
+	return unfolding acc(raw.Mem(ubuf), R50) in unfolding acc(sl.AbsSlice_Bytes(ubuf, 0, len(ubuf)), R51) in
+	let hop := path.BytesToIO_HF_Helper(ubuf, start, middle, end) in
+	hop == h.ToIO_HF()
+}
+
+ghost
+opaque
+requires  0 <= idx
+requires  acc(s.Mem(ubuf), R30)
+requires  acc(sl.AbsSlice_Bytes(ubuf, 0, len(ubuf)), R30)
+requires  idx < s.GetNumHops(ubuf)
+decreases
+pure func (s *Raw) CorrectlyDecodedHF_RawOffset(h path.HopField, idx int, ubuf []byte) bool {
+	return unfolding acc(s.Mem(ubuf), R31) in unfolding acc(s.Base.Mem(), R31) in
+	let hopOffset := MetaLen + s.NumINF*path.InfoLen + idx*path.HopLen in
+	hopOffset+path.HopLen <= len(s.Raw) &&
+	s.CorrectlyDecodedHF_Raw(h, hopOffset, hopOffset, hopOffset + path.HopLen, ubuf)
+}
+
+ghost
+opaque
+requires  acc(s.Mem(ubuf), R30)
+requires  acc(sl.AbsSlice_Bytes(ubuf, 0, len(ubuf)), R30)
+decreases
+pure func (s *Raw) CorrectlyDecodedHF_RawOffsetIndex(h path.HopField, ubuf []byte) bool {
+	return unfolding acc(s.Mem(ubuf), R31) in unfolding acc(s.Base.Mem(), R31) in
+	let idx := int(s.PathMeta.CurrHF) in
+	let hopOffset := MetaLen + s.NumINF*path.InfoLen + idx*path.HopLen in
+	let _ := Assuming(idx >= 0) in // Gobra can't figure this out from the type system at the moment
```

Comment:
Hopefully, this will not be a problem soon because I intend to fix the overflow checking of Gobra soon and to make it generate the necessary obligations here.

At any rate, I would like to avoid introduce assumptions in pure functions due to their "global" nature. I would rather have a precondition in this method requiring `s.PathMeta.CurrHF` to be non-negative and introducing an assumption in the caller (if it is a non-pure function) or a precondition (if it is a pure function). I think that opening a precedent where we introduce assumptions in the bodies of pure functions is a bad idea